File: FinSigth-Rest/app/schemas/goals.py
"""
Pydantic schemas for the goals API.

DB changes required (run manually):
  CREATE TABLE goal_investments (
      id         UUID PRIMARY KEY DEFAULT gen_random_uuid(),
      user_id    UUID NOT NULL REFERENCES users(id),
      goal_id    UUID NOT NULL REFERENCES user_goals(id) ON DELETE CASCADE,
      amount     DECIMAL(15,2) NOT NULL CHECK (amount > 0),
      date       DATE NOT NULL DEFAULT CURRENT_DATE,
      note       VARCHAR(500),
      created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
  );
  CREATE INDEX idx_goal_investments_user_goal ON goal_investments(user_id, goal_id);
"""

from typing import Any, Optional
from pydantic import BaseModel, field_validator


# ─────────────────────────────────────────────
# Request: generate plan
# ─────────────────────────────────────────────

class GoalRequest(BaseModel):
    goal_amount: float
    goal_months: int
    income_override: Optional[float] = None   # user-supplied income if estimate is wrong
    exclude_goal_id: Optional[str] = None     # set when adjusting an existing goal so it isn't counted as committed saving

    @field_validator('goal_amount')
    @classmethod
    def amount_must_be_positive(cls, v):
        if v <= 0:
            raise ValueError('goal_amount must be positive')
        return v

    @field_validator('goal_months')
    @classmethod
    def months_must_be_positive(cls, v):
        if v <= 0:
            raise ValueError('goal_months must be at least 1')
        return v

    @field_validator('income_override')
    @classmethod
    def income_override_positive(cls, v):
        if v is not None and v <= 0:
            raise ValueError('income_override must be positive')
        return v


# ─────────────────────────────────────────────
# Request: save a finalised plan
# ─────────────────────────────────────────────

class GoalSaveRequest(BaseModel):
    goal_name: str
    goal_amount: float
    goal_months: int
    required_monthly_saving: float
    monthly_income_used: float
    income_override: Optional[float] = None
    cluster_id: Optional[int] = None
    cluster_label: Optional[str] = None
    decisions: dict[str, Any]          # {category: {status, amount}}
    total_monthly_cutback: float
    baselines: dict[str, float] = {}   # {category: avg_monthly_spend at plan creation}


# ─────────────────────────────────────────────
# Inner response models
# ─────────────────────────────────────────────

class CategoryCutback(BaseModel):
    category: str
    current_monthly_spend: float
    recommended_monthly_spend: float
    monthly_saving: float                  # max saving available
    peer_avg_monthly_spend: float


class CategoryDrift(BaseModel):
    category: str
    plan_baseline: float    # avg monthly spend stored at goal creation
    current_avg: float      # avg monthly spend in last 3 months
    drift_pct: float        # (current - baseline) / baseline * 100; positive = spending more


# ─────────────────────────────────────────────
# Response: generate plan
# ─────────────────────────────────────────────

class GoalResponse(BaseModel):
    cluster_id: int
    cluster_label: str
    monthly_income_estimate: float
    current_monthly_saving: float          # raw saving from transactions
    committed_monthly_saving: float        # already locked by other saved goals
    available_monthly_saving: float        # current - committed (what's free for this goal)
    required_monthly_saving: float         # goal_amount / goal_months
    shortfall: float                       # required - available (0 if already covered)
    achievable_from_overspend: bool
    covered_monthly_saving: float
    recommendations: list[CategoryCutback]
    message: str
    investment_insight: Optional[str] = None   # positive opportunity flag, not a cutback


class MonthlyContribution(BaseModel):
    month: str               # 'YYYY-MM'
    net_surplus: float       # smoothed_salary - net_expenses (informational)
    total_invested: float    # Investment debits this month (informational)
    contribution: float      # plan-adherence: sum of min(max(0, baseline_cat - actual_cat), cutback) per category
    target: float            # goal's required_monthly_saving
    status: str              # 'ahead' | 'on_track' | 'behind'

class GoalTracking(BaseModel):
    monthly: list[MonthlyContribution]
    cumulative_contribution: float
    cumulative_target: float          # months_elapsed × required_monthly_saving
    months_elapsed: int
    progress_pct: float               # cumulative_contribution / goal_amount × 100, capped at 100
    overall_status: str               # 'ahead' | 'on_track' | 'behind' | 'not_started'
    avg_monthly_contribution: float
    projected_months_to_goal: Optional[float]  # None if no contribution yet


# ─────────────────────────────────────────────
# Response: saved goal (with live status)
# ─────────────────────────────────────────────

class SavedGoal(BaseModel):
    id: str
    goal_name: str
    goal_amount: float
    goal_months: int
    required_monthly_saving: float
    monthly_income_used: float
    income_override: Optional[float]
    cluster_id: Optional[int]
    cluster_label: Optional[str]
    decisions: dict[str, Any]
    total_monthly_cutback: float
    # Existing-savings fields:
    accumulated_savings_at_creation: float = 0.0   # snapshot of Investments total at save time
    count_existing_savings: bool = False            # user toggle: count saved pot toward goal
    created_at: str
    baselines: dict[str, float] = {}               # avg monthly spend per category at creation
    # Computed at fetch time from live transactions:
    coverage_amount: float     # how much of required is covered by available saving
    coverage_percent: int      # 0-100
    status: str                # 'on_track' | 'at_risk' | 'off_track'
    tracking: Optional['GoalTracking'] = None
    spend_drift: list['CategoryDrift'] = []        # categories with ≥20% spend shift since creation
    current_month: Optional[str] = None            # 'YYYY-MM' of the most recent data month
    current_month_spend: dict[str, float] = {}     # actual spend per category in current_month
    # Goal-tagged investments (from goal_investments table):
    tagged_investments: list['GoalInvestment'] = []
    total_tagged_investment: float = 0.0           # sum of all tagged investment amounts


class ToggleExistingSavingsRequest(BaseModel):
    count_existing_savings: bool


# ─────────────────────────────────────────────
# Goal investments: manual tagged savings
# ─────────────────────────────────────────────

class GoalInvestmentRequest(BaseModel):
    amount: float
    date: str                       # YYYY-MM-DD
    note: Optional[str] = None

    @field_validator('amount')
    @classmethod
    def amount_positive(cls, v):
        if v <= 0:
            raise ValueError('amount must be positive')
        return v


class GoalInvestment(BaseModel):
    id: str
    goal_id: str
    amount: float
    date: str                       # YYYY-MM-DD
    note: Optional[str] = None
    created_at: str


class SavedGoalListResponse(BaseModel):
    goals: list[SavedGoal]
    current_monthly_saving: float       # live estimate: income - avg_spend
    monthly_income_estimate: float      # median credit income used for the saving calculation
    avg_monthly_spend: float            # average debit spend used for the saving calculation
    total_required_monthly_saving: float
    total_committed_cutback: float      # sum of accepted cutbacks across all goals

File: FinSight-Web/src/models/goals.ts
// ─── Requests ────────────────────────────────────────────────────────────────

export interface GoalRequest {
  goal_name: string
  goal_amount: number
  goal_months: number
  income_override?: number  // user-supplied if estimated income is wrong
}

export interface GoalSaveRequest {
  goal_name: string
  goal_amount: number
  goal_months: number
  required_monthly_saving: number
  monthly_income_used: number
  income_override?: number
  cluster_id?: number
  cluster_label?: string
  decisions: Record<string, RecommendationDecision>
  total_monthly_cutback: number
  baselines: Record<string, number>   // { category: avg_monthly_spend } at plan creation
}

// ─── Plan response (POST /goals) ──────────────────────────────────────────────

export interface CategoryCutback {
  category: string
  current_monthly_spend: number
  recommended_monthly_spend: number
  monthly_saving: number           // max saving available from this category
  peer_avg_monthly_spend: number
}

export interface GoalResponse {
  cluster_id: number
  cluster_label: string
  monthly_income_estimate: number
  current_monthly_saving: number           // raw saving from transactions
  committed_monthly_saving: number         // locked by other saved goals
  available_monthly_saving: number         // current - committed
  required_monthly_saving: number          // goal_amount / goal_months
  shortfall: number                        // required - available (0 if on track)
  achievable_from_overspend: boolean
  covered_monthly_saving: number
  recommendations: CategoryCutback[]
  message: string
  investment_insight: string | null   // positive opportunity flag; null when already on track
}

// ─── Spend drift ──────────────────────────────────────────────────────────────

export interface CategoryDrift {
  category: string
  plan_baseline: number   // avg monthly spend stored at goal creation
  current_avg: number     // avg monthly spend in last 3 months
  drift_pct: number       // (current - baseline) / baseline * 100; positive = spending more
}

// ─── Recommendation decisions ─────────────────────────────────────────────────

export type RecommendationStatus = 'accepted' | 'modified' | 'skipped'

export interface RecommendationDecision {
  status: RecommendationStatus
  amount: number
}

// ─── Goal investments ─────────────────────────────────────────────────────────

export interface GoalInvestment {
  id: string
  goal_id: string
  amount: number
  date: string          // YYYY-MM-DD
  note?: string
  created_at: string
}

// ─── Goal tracking ────────────────────────────────────────────────────────────

export interface MonthlyContribution {
  month: string             // 'YYYY-MM'
  net_surplus: number       // informational: salary - expenses
  total_invested: number    // informational: investment debits
  contribution: number      // plan-adherence: ₹ of spending plan actually followed
  target: number
  status: 'ahead' | 'on_track' | 'behind' | 'gap'
}

export interface GoalTracking {
  monthly: MonthlyContribution[]
  cumulative_contribution: number
  cumulative_target: number
  months_elapsed: number
  progress_pct: number
  overall_status: 'ahead' | 'on_track' | 'behind' | 'not_started'
  avg_monthly_contribution: number
  projected_months_to_goal: number | null
}

// ─── Saved goal (GET /goals) ──────────────────────────────────────────────────

export interface SavedGoal {
  id: string
  goal_name: string
  goal_amount: number
  goal_months: number
  required_monthly_saving: number
  monthly_income_used: number
  income_override?: number
  cluster_id?: number
  cluster_label?: string
  decisions: Record<string, RecommendationDecision>
  total_monthly_cutback: number
  // Existing-savings fields:
  accumulated_savings_at_creation: number  // snapshot of Investments total at save time
  count_existing_savings: boolean          // user toggle: count saved pot toward goal
  created_at: string        // YYYY-MM-DD
  baselines: Record<string, number>   // avg monthly spend per category at creation
  // Computed at fetch time:
  coverage_amount: number   // how much of required is covered from available saving
  coverage_percent: number  // 0-100
  status: 'on_track' | 'at_risk' | 'off_track'
  tracking?: GoalTracking
  spend_drift?: CategoryDrift[]       // categories with ≥20% spend shift since creation
  current_month?: string              // 'YYYY-MM' of the most recent data month
  current_month_spend?: Record<string, number>  // actual spend per category in current_month
  // Goal-tagged investments (from goal_investments table):
  tagged_investments?: GoalInvestment[]
  total_tagged_investment?: number    // sum of all tagged investment amounts
}

export interface SavedGoalListResponse {
  goals: SavedGoal[]
  current_monthly_saving: number        // income - avg_spend (what's free each month)
  monthly_income_estimate: number       // median credit income used for the calculation
  avg_monthly_spend: number             // average debit spend used for the calculation
  total_required_monthly_saving: number
  total_committed_cutback: number
}
